File: src/frontend/src/utils/workflowStateMachine.ts
import { StoryStatus, TaskStatus } from '@/types/enums';

export const storyTransitions: Record<StoryStatus, StoryStatus[]> = {
    [StoryStatus.Backlog]: [StoryStatus.Ready],
    [StoryStatus.Ready]: [StoryStatus.InProgress],
    [StoryStatus.InProgress]: [StoryStatus.InReview],
    [StoryStatus.InReview]: [StoryStatus.QA, StoryStatus.InProgress],
    [StoryStatus.QA]: [StoryStatus.Done, StoryStatus.InProgress],
    [StoryStatus.Done]: [StoryStatus.Closed],
    [StoryStatus.Closed]: [],
};

export const taskTransitions: Record<TaskStatus, TaskStatus[]> = {
    [TaskStatus.ToDo]: [TaskStatus.InProgress],
    [TaskStatus.InProgress]: [TaskStatus.InReview],
    [TaskStatus.InReview]: [TaskStatus.InProgress, TaskStatus.Done],
    [TaskStatus.Done]: [],
};

export function getValidTransitions(
    entityType: 'Story' | 'Task',
    currentStatus: string
): string[] {
    if (entityType === 'Story') {
        return storyTransitions[currentStatus as StoryStatus] ?? [];
    }
    return taskTransitions[currentStatus as TaskStatus] ?? [];
}

export function isValidTransition(
    entityType: 'Story' | 'Task',
    from: string,
    to: string
): boolean {
    const valid = getValidTransitions(entityType, from);
    return valid.includes(to);
}

File: src/backend/WorkService/WorkService.Domain/Helpers/WorkflowStateMachine.cs
namespace WorkService.Domain.Helpers;

public static class WorkflowStateMachine
{
    private static readonly Dictionary<string, HashSet<string>> StoryTransitions = new()
    {
        ["Backlog"] = new() { "Ready" },
        ["Ready"] = new() { "InProgress" },
        ["InProgress"] = new() { "InReview" },
        ["InReview"] = new() { "InProgress", "QA" },
        ["QA"] = new() { "InProgress", "Done" },
        ["Done"] = new() { "Closed" },
        ["Closed"] = new()
    };

    private static readonly Dictionary<string, HashSet<string>> TaskTransitions = new()
    {
        ["ToDo"] = new() { "InProgress" },
        ["InProgress"] = new() { "InReview" },
        ["InReview"] = new() { "InProgress", "Done" },
        ["Done"] = new()
    };

    public static bool IsValidStoryTransition(string from, string to)
        => StoryTransitions.TryGetValue(from, out var targets) && targets.Contains(to);

    public static bool IsValidTaskTransition(string from, string to)
        => TaskTransitions.TryGetValue(from, out var targets) && targets.Contains(to);

    public static IReadOnlyDictionary<string, HashSet<string>> GetStoryTransitions() => StoryTransitions;
    public static IReadOnlyDictionary<string, HashSet<string>> GetTaskTransitions() => TaskTransitions;
}
